pkg/controller/controller.go

```diff
@@ -690,6 +695,43 @@ func (s *Service) ControllerUnpublishVolume(ctx context.Context, req *csi.Contro
 	return nil, status.Errorf(codes.InvalidArgument, "can't figure out protocol")
 }
 
+// GetServiceTag returns the service tag associated with an appliance
+func GetServiceTag(ctx context.Context, req *csi.CreateVolumeRequest, arr *array.PowerStoreArray, volID string, protocol string) string {
+	var ap gopowerstore.ApplianceInstance
+	var vol gopowerstore.Volume
+	var f gopowerstore.FileSystem
+	var nas gopowerstore.NAS
+	var applianceName string
+
+	if applianceID, ok := (req.Parameters)["appliance_id"]; ok {
+		ap, _ = arr.Client.GetAppliance(ctx, applianceID)
+	} else {
+		if protocol != "nfs" {
+			vol, _ = arr.Client.GetVolume(ctx, volID)
+			if vol.ApplianceID == "" {
```

Comment:
Please add nil check for all structs vol, f, ap, nas - before accessing the member variables.